### PR TITLE
fix(metrics): Fetch project config when metrics are received

### PR DESCRIPTION
**Pull request by jjbayer**
Metrics that arrive through the global metrics endpoint do not currently trigger a refetch of the project config. If a single processing relay only receives metrics traffic (no envelopes) for a specific project, metrics might get stuck in the pre-aggregator a.k.a. metrics buffer.

ref: https://github.com/getsentry/relay/issues/2967.

### CODE REVIEW

File: relay-server/src/services/project_cache.rs [hunk 1]
```diff
@@ -859,17 +859,19 @@ impl ProjectCacheBroker {
...
-            .merge_buckets(
-                aggregator,
-                outcome_aggregator,
-                envelope_processor,
-                message.buckets(),
-            );
+
+        let project = self.get_or_create_project(message.project_key());
+        project.prefetch(project_cache, false);
+        project.merge_buckets(
```
Comment (jjbayer) [inline relay-server/src/services/project_cache.rs:869]: Without this change, the added integration test fails.

File: relay-server/src/services/project.rs [hunk 2]
```diff
@@ -711,6 +711,10 @@ impl Project {
                         relay_log::debug!("sending metrics straight to aggregator");
                         let state = state.clone();
 
+                        // TODO: When the state is present but expired, we should send buckets
+                        // to the metrics buffer instead. In practice, the project state should be
+                        // refreshed at the time when the buckets emerge from the aggregator though.
```
Comment (jjbayer) [inline relay-server/src/services/project.rs:716]: I think we should do this, but I'll leave it for a follow-up.

File: tests/integration/test_metrics.py [hunk 3]
```diff
@@ -241,6 +241,36 @@ def test_global_metrics(mini_sentry, relay):
...
+            "name": "c:transactions/foo@none",
+            "value": 17.0,
+            "type": "c",
+        }
+    ]
+    relay.send_metrics_batch(
+        {"buckets": {public_key: metrics}},
+    )
+
+    envelope = mini_sentry.captured_events.get(timeout=3)
```
Comment (jjbayer) [inline tests/integration/test_metrics.py:265]: This `.get()` fails without the `prefetch`.

**Review by Dav1dde — APPROVED** [review]

File: relay-server/src/services/project_cache.rs [hunk 4]
```diff
@@ -859,17 +859,19 @@ impl ProjectCacheBroker {
...
-        // Only keep if we have an aggregator, otherwise drop because we know that we were disabled.
-        self.get_or_create_project(message.project_key())
-            .merge_buckets(
-                aggregator,
-                outcome_aggregator,
-                envelope_processor,
-                message.buckets(),
-            );
+
+        let project = self.get_or_create_project(message.project_key());
```
Comment (Dav1dde) [inline relay-server/src/services/project_cache.rs:867]: I wonder if we should just always prefetch in `get_or_create_project`. I found the same "bug" for metric metadata as well.

**Review by TBS1996 — APPROVED** [review]

File: relay-server/src/services/project_cache.rs (hunk 4 above)
Comment (jjbayer) [reply relay-server/src/services/project_cache.rs:867]: @Dav1dde I would even go further and unify the multiple state enums we have for project configs at the moment:

* `GetOrFetch` - either a cached project config or a channel
* `State` - either a cached project config or a metrics buffer
* `ExpiryState` - either `Updated`, `Stale`, or `Expired`

We might be able to make a single enum out of these, and give them an interface that makes it impossible to access a project without prefetching.

**Pull request merged**